src/Val.h

```diff
@@ -1497,20 +1496,20 @@ class RecordVal final : public Val, public notifier::detail::Modifiable {
     const std::vector<bool>& is_managed;
 };
 
-class EnumVal final : public detail::IntValImplementation {
+class EnumVal final : public detail::UnsignedValImplementation {
```

Comment:
Hmm, not comfortable doing this. It's breaking the API.

Any user calling `AsInt()` on an `ValPtr` today is valid (one example is in EventTrace.cc, another in logging/Manager.cc and possibly some in external plugins). After this change `AsInt()` will cast to the wrong underlying implementation. Things may still happen to work out, but it's confusing, no? Also, would expect UBSAN to be unhappy (can label the PR fullci and re-push to see if it would fail).

Can we keep the IntValImplementation and cap enum values a 0x7FFFFFFF...?

Alternatively, as mentioned in the ticket, sizeof enum/port doesn't seem to be documented, so IMO, deprecation could be another direction to take.

Reply:
It's fair to say that it may break some cases. But isn't it just as confusing that a value that is [enforced to be non-negative](https://github.com/zeek/zeek/blob/2907d9feeec4c107f7b728ff16ba1107287bd285/src/parse.y#L1144) is represented as a signed value? It makes a confusing case where an internal API can allow negative values technically, even though it's certainly not the intent. It's already assumed `EnumVal` will always be non-negative; making that concrete seems like the *less* confusing option.

I pushed the small test change and will see what fullci says here, but it seems like those cases can be easily fixed. More details on what to fix/how to fix it could go in the breaking changes section. If that's wrong then this can get scrapped easily.

Wouldn't it also be wrong to grab the enum value `AsInt` and then actually get a negative value? We already assume it won't be negative (if it's `-1` for example, then [`Lookup`'s result](https://github.com/zeek/zeek/blob/ba91de59b0b42126c2603eac9a27aa7cdc5aea75/src/Type.h#L831-L833) will be completely invalid). Wouldn't having a possibly-negative type introduce other subtle bugs? Wouldn't making `EnumVal` always unsigned solidify assumptions that were already made?

I don't really like deprecating it, if only because `SizeVal` is a [virtual method in `Val`](https://github.com/zeek/zeek/blob/ba91de59b0b42126c2603eac9a27aa7cdc5aea75/src/Val.h#L129-L131) so it makes sense to me to just implement it for all of them. But that's a pretty weak argument on my part

For what it's worth, the fix without the breaking change could just copy [int's version](https://github.com/zeek/zeek/blob/2907d9feeec4c107f7b728ff16ba1107287bd285/src/Val.cc#L216-L220) - the operator is also referred to as the "absolute value operator" for other types in the docs, so it makes more sense to me that it'd return an absolute value for enums, too.

Oh, also, regardless we should cap enum values at the max int value IMO. Or at least within an enum type. It still seems confusing to use `AsInt`, to me, since it can't be signed by design after that. But it'd at least help without introducing a breaking change (or as much of a breaking change, it could still potentially break something, technically). That still seems like it's trying to address the problem without actually addressing it, though

tl;dr: I'd be happy to just close this and make the simple change that's the fix for the problem without a bigger breaking change. I just also think this is an opportunity to make `EnumVal` more intuitive and less bug prone for users. I'd personally prefer making `EnumVal` express its non-negative reality.

Reply:
> I just also think this is an opportunity to make EnumVal more intuitive and less bug prone for users. I'd personally prefer making EnumVal express its non-negative reality.

I totally agree with your statements around making the underlying implementation unsigned. I'm solely coming from the angle that there are existing users calling `AsInt()` on a `ValPtr` and they now invoke undefined behavior (see the downcasting errors in the ubsan CI job for internal ones). And also, that we try to avoid changing the API in the public `zeek::` namespace without a deprecation cycle.

> but it seems like those cases can be easily fixed.

EnumVal lives in the public `zeek::` namespace (and is an established/old API). Cases and callers might live in code we do not control, have no access or even knowledge of. We usually try to deprecate and raise compilation 
warnings (via e.g. `[[deprecated]]`) for changes to the API for a whole LTS cycle (in this case it would be throughout 8.0) so users have time to adapt. I'm just not sure how that would/could look and whether it's worth it - maybe deprecating `AsInt()` on Val and implementing it just for `IntVal` (but not sure this works out or is great).

Reply:
Well, that's fair enough. It really doesn't feel like anything is deprecated, so deprecating `AsInt()` for generic values seems like a pretty heavy-handed approach to get this one change in. I'll just close this and do the easy way that just directly fixes that issue. I was just trying to sneak in something to make this case cleaner while also solidifying some assumptions

For what it's worth, I still really like this change and would like to find a way to get the underlying enum value as unsigned 🙊  but I'll keep that on the back burner a little